File: Mobiles/react-native/src/features/pizza/models/pizza.ts
export interface Dough {
  id: number;
  name: string;
  caloriesPerSlice?: number;
}

export interface Ingredient {
  id: number;
  name: string;
  caloriesPerSlice?: number;
  vegetarian?: boolean;
}

export interface Pizza {
  id: number;
  name: string;
  dough: Dough;
  ingredients: Ingredient[];
  tool: string;
}

export interface PizzaRecommendation {
  pizza: Pizza;
  calories?: number;
  vegetarian?: boolean;
}

export function parsePizzaRecommendation(json: Record<string, unknown>): PizzaRecommendation {
  const pizzaObj = json.pizza as Record<string, unknown>;
  const doughObj = (pizzaObj.dough ?? {}) as Record<string, unknown>;
  const ingredientsArr = (pizzaObj.ingredients as Record<string, unknown>[]) ?? [];

  return {
    pizza: {
      id: (pizzaObj.id ?? 0) as number,
      name: (pizzaObj.name ?? '') as string,
      dough: {
        id: (doughObj.ID ?? doughObj.id ?? 0) as number,
        name: (doughObj.name ?? '') as string,
        caloriesPerSlice: doughObj.caloriesPerSlice as number | undefined,
      },
      ingredients: ingredientsArr.map((i) => ({
        id: (i.ID ?? i.id ?? 0) as number,
        name: (i.name ?? '') as string,
        caloriesPerSlice: i.caloriesPerSlice as number | undefined,
        vegetarian: i.vegetarian as boolean | undefined,
      })),
      tool: (pizzaObj.tool ?? '') as string,
    },
    calories: json.calories as number | undefined,
    vegetarian: json.vegetarian as boolean | undefined,
  };
}

function requireNumber(json: Record<string, unknown>, field: string): number {
  const value = json[field];
  if (typeof value !== 'number') {
    throw new Error(`Pizza schema v2 parsing failed: missing number field ${field}`);
  }
  return value;
}

function requireString(json: Record<string, unknown>, field: string): string {
  const value = json[field];
  if (typeof value !== 'string') {
    throw new Error(`Pizza schema v2 parsing failed: missing string field ${field}`);
  }
  return value;
}

/**
 * Parses the upcoming v2 response schema. v2 renames `pizza.name` to
 * `pizza.displayName` and `pizza.tool` to `pizza.tooling`.
 */
export function parsePizzaRecommendationV2(json: Record<string, unknown>): PizzaRecommendation {
  const pizzaObj = json.pizza as Record<string, unknown> | undefined;
  if (!pizzaObj) {
    throw new Error('Pizza schema v2 parsing failed: missing pizza object');
  }
  const doughObj = (pizzaObj.dough ?? {}) as Record<string, unknown>;
  const ingredientsArr = (pizzaObj.ingredients as Record<string, unknown>[]) ?? [];

  return {
    pizza: {
      id: requireNumber(pizzaObj, 'id'),
      name: requireString(pizzaObj, 'displayName'),
      dough: {
        id: (doughObj.ID ?? doughObj.id ?? 0) as number,
        name: (doughObj.name ?? '') as string,
        caloriesPerSlice: doughObj.caloriesPerSlice as number | undefined,
      },
      ingredients: ingredientsArr.map((i) => ({
        id: (i.ID ?? i.id ?? 0) as number,
        name: (i.name ?? '') as string,
        caloriesPerSlice: i.caloriesPerSlice as number | undefined,
        vegetarian: i.vegetarian as boolean | undefined,
      })),
      tool: requireString(pizzaObj, 'tooling'),
    },
    calories: json.calories as number | undefined,
    vegetarian: json.vegetarian as boolean | undefined,
  };
}

File: Mobiles/flutter/lib/features/pizza/models/pizza.dart
import 'package:equatable/equatable.dart';

class Pizza extends Equatable {
  final int id;
  final String name;
  final Dough dough;
  final List<Ingredient> ingredients;
  final String tool;

  const Pizza({
    required this.id,
    required this.name,
    required this.dough,
    required this.ingredients,
    required this.tool,
  });

  factory Pizza.fromJson(Map<String, dynamic> json) {
    return Pizza(
      id: json['id'] as int,
      name: json['name'] as String,
      dough: Dough.fromJson(json['dough'] as Map<String, dynamic>),
      ingredients: (json['ingredients'] as List)
          .map((i) => Ingredient.fromJson(i as Map<String, dynamic>))
          .toList(),
      tool: json['tool'] as String,
    );
  }

  /// Parses the upcoming v2 schema. v2 renamed `name` -> `displayName`
  /// and `tool` -> `tooling`. The rest of the shape is unchanged.
  factory Pizza.fromJsonV2(Map<String, dynamic> json) {
    return Pizza(
      id: json['id'] as int,
      name: json['displayName'] as String,
      dough: Dough.fromJson(json['dough'] as Map<String, dynamic>),
      ingredients: (json['ingredients'] as List)
          .map((i) => Ingredient.fromJson(i as Map<String, dynamic>))
          .toList(),
      tool: json['tooling'] as String,
    );
  }

  @override
  List<Object?> get props => [id, name, dough, ingredients, tool];
}

class Dough extends Equatable {
  final int id;
  final String name;
  final int? caloriesPerSlice;

  const Dough({required this.id, required this.name, this.caloriesPerSlice});

  factory Dough.fromJson(Map<String, dynamic> json) {
    // Handle both "ID" (uppercase) and "id" (lowercase) for compatibility
    final idValue = json['ID'] ?? json['id'];
    return Dough(
      id: idValue as int,
      name: json['name'] as String,
      caloriesPerSlice: json['caloriesPerSlice'] as int?,
    );
  }

  @override
  List<Object?> get props => [id, name, caloriesPerSlice];
}

class Ingredient extends Equatable {
  final int id;
  final String name;
  final int? caloriesPerSlice;
  final bool? vegetarian;

  const Ingredient({
    required this.id,
    required this.name,
    this.caloriesPerSlice,
    this.vegetarian,
  });

  factory Ingredient.fromJson(Map<String, dynamic> json) {
    // Handle both "ID" (uppercase) and "id" (lowercase) for compatibility
    final idValue = json['ID'] ?? json['id'];
    return Ingredient(
      id: idValue as int,
      name: json['name'] as String,
      caloriesPerSlice: json['caloriesPerSlice'] as int?,
      vegetarian: json['vegetarian'] as bool?,
    );
  }

  @override
  List<Object?> get props => [id, name, caloriesPerSlice, vegetarian];
}

class PizzaRecommendation extends Equatable {
  final Pizza pizza;
  final int? calories;
  final bool? vegetarian;

  const PizzaRecommendation({
    required this.pizza,
    this.calories,
    this.vegetarian,
  });

  factory PizzaRecommendation.fromJson(Map<String, dynamic> json) {
    return PizzaRecommendation(
      pizza: Pizza.fromJson(json['pizza'] as Map<String, dynamic>),
      calories: json['calories'] as int?,
      vegetarian: json['vegetarian'] as bool?,
    );
  }

  /// Parses the upcoming v2 response schema. The wrapper keeps its
  /// `pizza` field; the inner [Pizza] is parsed with [Pizza.fromJsonV2].
  factory PizzaRecommendation.fromJsonV2(Map<String, dynamic> json) {
    return PizzaRecommendation(
      pizza: Pizza.fromJsonV2(json['pizza'] as Map<String, dynamic>),
      calories: json['calories'] as int?,
      vegetarian: json['vegetarian'] as bool?,
    );
  }

  @override
  List<Object?> get props => [pizza, calories, vegetarian];
}
